cocos/renderer/gfx-gles2/GLES2Commands.cc

```diff
@@ -1899,10 +1899,149 @@ void GLES2CmdFuncExecuteCmds(GLES2Device* device, GLES2CmdPackage* cmd_package)
           } // if
         }
           
-        //TODO: implementation setting dynamic states.
-        
+        if(gpuPipelineState)
+        {
+            for(auto dynamicState : gpuPipelineState->dynamicStates)
+            {
+                switch (dynamicState)
+                {
+                case GFXDynamicState::VIEWPORT:
+                    if(cache->viewport != cmd->viewport)
+                    {
+                        cache->viewport = cmd->viewport;
+                        glViewport(cmd->viewport.left, cmd->viewport.top, cmd->viewport.width, cmd->viewport.height);
+                    }
+                    break;
+                case GFXDynamicState::SCISSOR:
+                    if(cache->scissor != cmd->scissor)
+                    {
+                        cache->scissor = cmd->scissor;
+                        glScissor(cmd->scissor.x, cmd->scissor.y, cmd->scissor.width, cmd->scissor.height);
+                    }
+                    break;
+                case GFXDynamicState::LINE_WIDTH:
+                    if(cache->rs.lineWidth != cmd->lineWidth)
+                    {
+                        cache->rs.lineWidth = cmd->lineWidth;
+                        glLineWidth(cmd->lineWidth);
+                    }
+                    break;
+                case GFXDynamicState::DEPTH_BIAS:
+                    if(cache->rs.depthBiasEnabled != cmd->depthBiasEnabled)
+                    {
+                        if(cmd->depthBiasEnabled)
+                          glEnable(GL_POLYGON_OFFSET_FILL);
+                        else
+                          glDisable(GL_POLYGON_OFFSET_FILL);
+
+                        cache->rs.depthBiasEnabled = cmd->depthBiasEnabled;
+                    }
+                    if((cache->rs.depthBias != cmd->depthBias.constant) ||
+                       (cache->rs.depthBiasSlop != cmd->depthBias.slope))
+                    {
+                        glPolygonOffset(cmd->depthBias.constant, cmd->depthBias.slope);
+                        cache->rs.depthBias = cmd->depthBias.constant;
+                        cache->rs.depthBiasSlop = cmd->depthBias.slope;
+                    }
+                    break;
+                case GFXDynamicState::BLEND_CONSTANTS:
+                    if ((cache->bs.blendColor.r != gpuPipelineState->bs.blendColor.r) ||
+                        (cache->bs.blendColor.g != gpuPipelineState->bs.blendColor.g) ||
+                        (cache->bs.blendColor.b != gpuPipelineState->bs.blendColor.b) ||
+                        (cache->bs.blendColor.a != gpuPipelineState->bs.blendColor.a)) 
```

Comment:
Can implement operator == and != in GFXColor to simplify the codes. And float comparison can not just use !=.

Reply:
got it.